### PR TITLE
Hive: Support connecting to multiple Hive-Catalog

### PR DESCRIPTION
**Pull request by szehon-ho**
_Note:   As the term Hive-Catalog is overloaded across, I'll use "Iceberg-HiveCatalog" to refer to Iceberg's HiveCatalog class and "HMS Catalog" to refer to Hive specific concept._

https://issues.apache.org/jira/browse/HIVE-18685 added support for HMS catalogs, ie different namespaces for dbs/tables in the metastore.

However, the Iceberg-HiveCatalog uses a global cache of HMS connections based on Hive-Metastore-URI.  This prevents different Iceberg-HiveCatalogs existing in same JVM from talking to different HMS Catalogs on same HMS.  See issue:  https://github.com/apache/iceberg/pull/5378.  This is important for example in Spark, where user may try to configure different SparkCatalog pointing to separate catalog on same HMS.   

This change takes advantage of https://github.com/apache/iceberg/pull/6698 to fix this.  It does two things:

- Add a 'hive-catalog'  flag on Iceberg-HiveCatalog (similar to "uri" and "warehouse"), which automatically sets this on HMS client to point to right HMS Catalog.  
- Adds the HMS client config "metastore.catalog.default" to the list of HMS Client cache keys.

From Spark point of view, user can configure 'spark.sql.catalog.$cat1.hive-catalog=foo', 'spark.sql.catalog.$cat2.hive-catalog=bar'.

Note; @RussellSpitzer pointed to me that setting  spark.sql.catalog.$myCatalog.hadoop.metastore.catalog.default=foo seems to work to configure each SparkCatalog.  But it is a bit clunky and undocumented, and hence the proposal to add 'hive_catalog' shorthand.

In either case, the second change is needed to allow different cached connections on same HMS to go to different HMS-catalogs.

### CODE REVIEW

**Review by pvary — APPROVED** [review]

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 1]
```diff
@@ -64,10 +64,15 @@ public class HiveCatalog extends BaseMetastoreCatalog implements SupportsNamespa
   public static final String LIST_ALL_TABLES = "list-all-tables";
   public static final String LIST_ALL_TABLES_DEFAULT = "false";
 
+  public static final String HMS_CATALOG = "hive-catalog";
```
Comment (pvary) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:67]: Do we have to document this?

Comment (szehon-ho) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:67]: Will do in follow-up pr, thanks

**Review by hililiwei — APPROVED** [review]
It looks good. I just have quick question, when we use multiple hive-catalog, we can perform complex queries between tables (even in different clusters) that are completely isolated by catalog, such as join\union, right?

**Comment by szehon-ho** [comment]
Thanks.  Yes, it depends on engine, I believe Spark has this support.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java (hunk 1 above)
Comment (szehon-ho) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java]: Actually decided not to add this extra config for now, as user can use spark.sql.catalog.$myCatalog.hadoop.metastore.catalog.default.  Can revisit this later, but for now no need to document.

**Comment by szehon-ho** [comment]
Test failed looks unrelated, kicking off again

**Comment by szehon-ho** [comment]
Last test cleanup is not related to failure (is just good to clear for next test).  Thanks @pvary @hililiwei  for review

**Pull request merged**